apex/helper/token_helper.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright (c) 2024 Dell Inc., or its subsidiaries. All Rights Reserved.
+
+Licensed under the Mozilla Public License Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://mozilla.org/MPL/2.0/
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package helper
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+)
+
+// GetNewToken retrieves a new token by running a Python script, reading from a named pipe,
+// and sending an API request.
+//
+// ctx: The context.Context object for the function.
+// Returns:
+// - string: The token from the response from the API request.
+// - error: An error if any occurred during the process.
+func GetNewToken(ctx context.Context) (string, error) {
+
+	// values taken from environmental variables
+	// Set the directory path
+	dirPath := os.Getenv("APEX_SAML_TOKEN_SCRIPT_DIR")
+	// Set the Python script path
+	scriptPath := os.Getenv("APEX_SAML_TOKEN_SCRIPT_NAME")
+
+	// Set the command to run the Python script
+	cmd := exec.Command("python", scriptPath)
+	cmd.Dir = dirPath
+
+	// Run the command and capture the output
+	_, err := cmd.CombinedOutput()
+	if err != nil {
+		tflog.Error(ctx, "Error running SAML get token:", map[string]interface{}{
+			"Error":   err,
+			"DirPath": dirPath,
+			"Script:": scriptPath,
+		})
+	}
+
+	tokenFileName := os.Getenv("APEX_SAML_TOKEN_OUTPUT_FILE_NAME")
```

Comment:
Looks saml assertion is generated by the python script and is written into a file. Could we include the python file in this repo? And why we separate the whole process (saml token in python and bearer token in golang)?

Reply:
@P-Cao we can discuss this.